File: Cpp/CXR_classify/storer.cpp
#include "storer.h"

#include <chrono>
#include <iostream>

Storer::Storer(ConfigHandler * configHandler, DatabaseHandler * dbHandler) : Runnable(configHandler, dbHandler)
{

}

void Storer::run()
{
    std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
    emit attemptUpdateText("Storing metadata");
    emit attemptUpdateProBarBounds(0, expected_num_files);

    std::string metadataTableName = configHandler->getTableName("metadata");
    std::string columnsInfoName = configHandler->getColumnsInfoName();

    // Create table if it isn't already there
    if (!dbHandler->tableExists(metadataTableName)) {
        dbHandler->addTableToDb(columnsInfoName, "elements", metadataTableName);
    }

    logger->info("Attempting to store DICOM metadata from DCMs in a folder to Postgres DB");

    emit attemptUpdateProBarValue(0);

    // Open the json with the list of elements we're interested in
    boost::property_tree::ptree columnsJson;
    boost::property_tree::read_json(columnsInfoName, columnsJson);
    boost::property_tree::ptree elements = columnsJson.get_child("elements");

    quint64 storeCount = 0;
    for (auto & p : std::filesystem::recursive_directory_iterator(configHandler->getUnpackFolderPath())) {
        if (p.path().extension() == ".dcm") {
            logger->debug("Storing: {}", p.path().string());
            pqxx::result result = dbHandler->executeQuery(dbHandler->connection, createSqlQuery(metadataTableName, elements, p.path().string()));
            storeCount++;
            emit attemptUpdateProBarValue(storeCount);
        }
    }

    logger->info("Done storing metadata");

    std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
    std::cout << "Time difference = " << (std::chrono::duration_cast<std::chrono::nanoseconds> (end - begin).count())/1000000000.0 << "[s]" << std::endl;
    emit attemptUpdateText("Done storing metadata");
    emit finished();
}

std::string Storer::createSqlQuery(std::string tableName, boost::property_tree::ptree elements, std::string filePath)
{
    // Read file
    DcmFileFormat file_format;
    OFCondition status = file_format.loadFile(filePath.c_str());

    // Go through the list of elements and try to read the value
    for (boost::property_tree::ptree::value_type & element : elements) {
        // Get tag and separate into its parts
        std::string tag = element.second.get<std::string>("tag");
        std::string groupNameString;
        std::string elementNumString;

        tag.erase(std::remove_if(tag.begin(), tag.end(), ::isspace), tag.end()); // Remove blank spaces
        std::istringstream tagStream(tag);
        getline(tagStream, groupNameString, ',');
        getline(tagStream, elementNumString, ',');

        uint16_t groupNameInt = std::stoul(groupNameString, nullptr, 16);
        uint16_t elementNumInt = std::stoul(elementNumString, nullptr, 16);

        DcmTagKey tagKey(groupNameInt, elementNumInt);
        OFString tagValue;
        file_format.getDataset()->findAndGetOFString(tagKey, tagValue);

        if (tagValue != "") {
            if (element.second.get<std::string>("db_datatype").find("INT") != std::string::npos) {
                element.second.put("value", tagValue.c_str());
            } else if (element.second.get<std::string>("db_datatype").find("CHAR") != std::string::npos) {
                std::string str(tagValue.c_str());
                element.second.put("value", "'" + str + "'");
            }
        } else {
            element.second.put("value", "NULL");
        }
    }

    std::vector<std::string> names{"file_name", "file_path"};
    std::vector<std::string> values{"'" + filePath.substr(filePath.find_last_of("/") + 1) + "'", "'" + filePath + "'"};

    for (boost::property_tree::ptree::value_type & element : elements) {
        if (!element.second.get<bool>("calculation_only")) {
            names.push_back(element.first);
            values.push_back(element.second.get<std::string>("value"));
        }
    }

    std::string sqlQuery = "INSERT INTO " + tableName + " (" + boost::algorithm::join(names, ", ") + ") VALUES (" + boost::algorithm::join(values, ", ") + ");";

    return sqlQuery;
}

File: Go/CxrClassify/storer/storer.go
package storer

import (
	"CxrClassify/configHandler"
	"CxrClassify/databaseHandler"
	"CxrClassify/runnable"
	"encoding/json"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/suyashkumar/dicom"
	"github.com/suyashkumar/dicom/pkg/tag"
)

type Storer struct {
	runnable.Runnable

	_ func() `constructor:"init"`

	MetadataTableName string
	ColumnsInfoName   string
}

func (s *Storer) init() {

}

func (s *Storer) Setup(configHandler *configHandler.ConfigHandler, databaseHandler *databaseHandler.DatabaseHandler) {
	s.SetupRunnable(configHandler, databaseHandler)
}

func (s Storer) Run() {
	// logging set up
	f, err := os.OpenFile("testlogfile", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
	if err != nil {
		log.Fatalf("error opening file: %v", err)
	}
	defer f.Close()
	log.SetOutput(f)

	s.AttemptUpdateText("Storing metadata")
	s.AttemptUpdateProBarBounds(0, 10)

	s.MetadataTableName = s.ConfigHandler.GetTableName("metadata")
	s.ColumnsInfoName = s.ConfigHandler.GetColumnsInfoName()

	if !s.DatabaseHandler.TableExists(s.MetadataTableName) {
		s.DatabaseHandler.AddTableToDb(s.ColumnsInfoName, "elements", s.MetadataTableName)
	}

	s.AttemptUpdateProBarValue(0)

	jsonFile, err := os.Open(s.ColumnsInfoName)
	if err != nil {
		s.AttemptUpdateText("DF")
	}
	defer jsonFile.Close()

	byteValue, _ := ioutil.ReadAll(jsonFile)

	var columnsJson map[string]interface{}
	json.Unmarshal([]byte(byteValue), &columnsJson)

	elements := columnsJson["elements"].(map[string]interface{})

	var storeCount = 0
	err = filepath.Walk(s.ConfigHandler.GetUnpackFolderPath(),
		func(path string, info os.FileInfo, err error) error {

			if err != nil {
				return err
			}

			fileExtension := filepath.Ext(path)

			if fileExtension == ".dcm" {
				dataset, _ := dicom.ParseFile(path, nil)

				for k, v := range elements {
					tagString := v.(map[string]interface{})["tag"].(string)
					tagSlice := strings.Split(tagString, ",")

					groupNameString := strings.Replace(strings.TrimSpace(tagSlice[0]), "0x", "", -1)
					elementNumString := strings.Replace(strings.TrimSpace(tagSlice[1]), "0x", "", -1)

					groupNameInt, _ := strconv.ParseUint(groupNameString, 16, 64)
					elementNumInt, _ := strconv.ParseUint(elementNumString, 16, 64)

					_, err := dataset.FindElementByTag(tag.Tag{Group: uint16(groupNameInt), Element: uint16(elementNumInt)})

					if err == nil {
						dbDatatype := v.(map[string]interface{})["db_datatype"].(string)
						elem, _ := dataset.FindElementByTag(tag.Tag{Group: uint16(groupNameInt), Element: uint16(elementNumInt)})
						if strings.Contains(dbDatatype, "INT") {
							value := elem.Value.String()
							elements[k].(map[string]interface{})["value"] = elem.Value.String()[1 : len(value)-1]
						} else if strings.Contains(dbDatatype, "CHAR") {
							value := elem.Value.String()
							elements[k].(map[string]interface{})["value"] = "'" + elem.Value.String()[1:len(value)-1] + "'"
						}
					} else {
						elements[k].(map[string]interface{})["value"] = "NULL"
					}
				}

				names := []string{"file_name", "file_path"}
				values := []string{"'" + filepath.Base(path) + "'", "'" + path + "'"}

				for k, v := range elements {
					if !v.(map[string]interface{})["calculation_only"].(bool) {
						names = append(names, k)
						values = append(values, v.(map[string]interface{})["value"].(string))
					}
				}

				sqlQuery := "INSERT INTO " + s.MetadataTableName + " (" + strings.Join(names, ", ") + ") VALUES (" + strings.Join(values, ", ") + ");"

				s.DatabaseHandler.ExecuteQuery(s.DatabaseHandler.Connection, sqlQuery)

				storeCount++

				s.AttemptUpdateProBarValue(storeCount)
			}

			return nil
		})
	if err != nil {
		log.Println("FAIL")
	}

	// sqlQuery := elements["patient_orientation"].(map[string]interface{})["db_datatype"].(string)

	// d.ExecuteQuery(d.Connection, sqlQuery)

	s.AttemptUpdateText("Done storing metadata")
	s.Finished()
}

// func (s Storer) CreateSqlQuery(tableName string, elements) {
